### PR TITLE
Use built-in validation for altitude

### PR DESCRIPTION
**Pull request by agners**
This improves the error location and message while being still backwards
compatible.

# What does this implement/fix? 

Quick description 

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):**

- Noted in: https://github.com/esphome/issues/issues/1840

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>
  
# Test Environment

- [x] ESP32
- [ ] ESP8266
- [ ] Windows
- [ ] Mac OS
- [x] Linux

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: scd30
    co2:
      id: sensirion
      name: "Sensirion CO2"
      retain: False
      accuracy_decimals: 1
    temperature:
      name: "Sensirion Temperature"
      retain: False
      accuracy_decimals: 2
    humidity:
      name: "Sensirion Humidity"
      retain: False
      accuracy_decimals: 1
    altitude_compensation: 500m
    address: 0x61
    update_interval: 60s
```

# Explain your changes

Describe your changes here to communicate to the maintainers **why we should accept this pull request**.
Very important to fill if no issue linked

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/scd30/sensor.py [hunk 1]
```diff
@@ -47,7 +42,7 @@ def remove_altitude_suffix(value):
             ),
             cv.Optional(CONF_AUTOMATIC_SELF_CALIBRATION, default=True): cv.boolean,
             cv.Optional(CONF_ALTITUDE_COMPENSATION): cv.All(
-                remove_altitude_suffix,
+                cv.float_with_unit("altitude", "(m|m a.s.l.|m MAMSL|m MAML)"),
```
Comment (jesserockz) [inline esphome/components/scd30/sensor.py:45]: `MAML` was not valid in the original regex.

Also does `MAMSL` and `MASL` need the `m` at the start when the first `M` is meters? (not sure why the original regex allowed that)

```suggestion
                cv.float_with_unit("altitude", "(m|m a.s.l.|m MAMSL|m MASL)"),
```

Comment (agners) [reply esphome/components/scd30/sensor.py:45]: Right!

File: esphome/components/scd30/sensor.py [hunk 2]
```diff
@@ -47,7 +42,7 @@ def remove_altitude_suffix(value):
             ),
             cv.Optional(CONF_AUTOMATIC_SELF_CALIBRATION, default=True): cv.boolean,
             cv.Optional(CONF_ALTITUDE_COMPENSATION): cv.All(
-                remove_altitude_suffix,
+                cv.float_with_unit("altitude", "(m|m a.s.l.|m MAMSL|m MASL)"),
```
Comment (jesserockz) [inline esphome/components/scd30/sensor.py:45]: While the regex allowed the `m`, I don't think it should be there - `meters Meters Above Sea Level` doesnt make sense
```suggestion
                cv.float_with_unit("altitude", "(m|m a.s.l.|MAMSL|MASL)"),
```

Comment (agners) [reply esphome/components/scd30/sensor.py:45]: good point!

**Review by jesserockz — APPROVED** [review]
Thanks

**Pull request merged**